src/components/structures/MessagePanel.js

```diff
@@ -0,0 +1,264 @@
+     */
+    checkFillState: function() {
+        if (this.refs.scrollPanel) {
+            this.refs.scrollPanel.checkFillState();
+        }
+    },
+
+    render: function() {
+        var ScrollPanel = sdk.getComponent("structures.ScrollPanel");
+        return (
+            <ScrollPanel ref="scrollPanel" className="mx_RoomView_messagePanel"
+                    onScroll={ this.props.onScroll } 
+                    onFillRequest={ this.props.onFillRequest }
+                    style={ this.props.hidden ? { display: 'none' } : {} }
+                    stickyBottom={ this.props.stickyBottom }>
+                {this._getEventTiles()}
+            </ScrollPanel>
+        );
+    },
+
+    _getEventTiles: function() {
+        var DateSeparator = sdk.getComponent('messages.DateSeparator');
+        var EventTile = sdk.getComponent('rooms.EventTile');
+
+        var ret = [];
+
+        var prevEvent = null; // the last event we showed
+        var ghostIndex;
+        var readMarkerIndex;
+        for (var i = 0; i < this.props.events.length; i++) {
+            var mxEv = this.props.events[i];
+
+            if (!EventTile.haveTileForEvent(mxEv)) {
+                continue;
+            }
+
+            if (this.props.isConferenceUser && mxEv.getType() === "m.room.member") {
+                if (this.props.isConferenceUser(mxEv.getSender()) ||
+                        this.props.isConferenceUser(mxEv.getStateKey())) {
+                    continue; // suppress conf user join/parts
+                }
+            }
+
+            // now we've decided whether or not to show this message,
+            // add the read up to marker if appropriate
+            // doing this here means we implicitly do not show the marker
+            // if it's at the bottom
+            // NB. it would be better to decide where the read marker was going
+            // when the state changed rather than here in the render method, but
+            // this is where we decide what messages we show so it's the only
+            // place we know whether we're at the bottom or not.
+            var mxEvSender = mxEv.sender ? mxEv.sender.userId : null;
+            if (prevEvent && prevEvent.getId() == this.props.readMarkerEventId) {
+                // suppress the read marker if the next event is sent by us; this
+                // is a nonsensical and temporary situation caused by the delay between
+                // us sending a message and receiving the synthesized receipt.
+                if (mxEvSender != this.props.ourUserId) {
+                    var hr;
+                    hr = (
+                        <hr className="mx_RoomView_myReadMarker"
+                            style={{opacity: 1, width: '99%'}} 
+                        />);
+                    readMarkerIndex = ret.length;
+                    ret.push(
+                        <li key="_readupto"
+                                className="mx_RoomView_myReadMarker_container">
```

Comment:
There doesn't seem to be a good way of making indentation work for attributes and nested elements. Have tweaked it anyway.